<a href="https://colab.research.google.com/github/akramlodi/Hate-Speech-Classification/blob/main/Hate_classification_using_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Hate Classification using Naive Bayes Algorithm

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
from pathlib import Path
import pandas as pd
import numpy as np
import re

df = pd.read_csv('sample_data/labeled_data.csv')

df['label'] = df.apply(lambda row: 'offensive' if row['hate_speech'] > 0 or row['offensive_language'] > 0 else 'non-offensive', axis=1)


import re
def clean_text(text):
    text = re.sub(r'http\S+', '', text)  # to Remove URLs
    text = re.sub(r'@\w+', '', text)     # to Remove mentions
    text = re.sub(r'#\w+', '', text)     # to Remove hashtags
    text = re.sub(r'\d+', '', text)      # to Remove numbers
    text = text.lower()                  # to Convert to lowercase
    text = re.sub(r'\W+', ' ', text)     # to Remove special characters
    text = re.sub(r'\s+', ' ', text)     # Remove extra spaces
    text = text.strip()                  # Remove leading/trailing spaces
    text = re.sub(r'rt', '', text)       # Remove retweets
    text = re.sub(r'\n', '', text)       # Remove newline characters

    return text



# Apply text cleaning
df['cleaned_tweet'] = df['tweet'].apply(clean_text)

# Select relevant columns
df = df[['cleaned_tweet', 'label']]
df.head()

cleaned_tweet          label
0   as a woman you shouldn t complain about clean...  non-offensive
1   boy dats cold tyga dwn bad for cuffin dat hoe...      offensive
2   dawg  you ever fuck a bitch and she sta to cr...      offensive
3                             she look like a tranny      offensive
4   the shit you hear about me might be true or i...      offensive

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Vectorize the text data
vectorizer = TfidfVectorizer(max_features=5000)
X = vectorizer.fit_transform(df['cleaned_tweet'])
y = df['label']


In [ ]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

# Initialize the Multinomial Naive Bayes model
nb_model = MultinomialNB()

# Train the model
nb_model.fit(X_train, y_train)

# Predict on the test data
y_pred = nb_model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print('Classification Report:')
print(report)


Accuracy: 0.8973169255598143
Classification Report:
               precision    recall  f1-score   support

non-offensive       0.77      0.16      0.26       573
    offensive       0.90      0.99      0.94      4384

     accuracy                           0.90      4957
    macro avg       0.84      0.58      0.60      4957
 weighted avg       0.89      0.90      0.87      4957



#Result and Conclusion
The Naive Bayes model achieved an overall accuracy of 90%. However, it performed poorly on the "non-offensive" class with a precision of 0.77 and a very low recall of 0.16, resulting in a low F1-score of 0.26. For the "offensive" class, the model performed well with a precision of 0.90, recall of 0.99, and an F1-score of 0.94. The significant class imbalance affected the performance, particularly for the "non-offensive" class. Addressing the class imbalance through techniques such as resampling, using class weights, or generating synthetic data could improve model performance.
